pennylane/templates/repeat.py

```diff
@@ -0,0 +1,211 @@
+"""
+# pylint: disable-msg=too-many-branches,too-many-arguments,protected-access
+from pennylane.templates.decorator import template
+
+###################
+
+
+@template
+def repeat(circuit, depth, *args, **kwargs):
+    r"""Repeatedly applies a circuit containing quantum gates or templates.
+
+    Args:
+        circuit (function): A function that applies the quantum gates/templates being repeated.
+        depth (int): The number of times the circuit is repeatedly applied.
+        *args: Dynamic parameters that are passed into ``circuit`` each time it is
+               repeated (see UsageDetails for more information). Each dynamic argument
+               must be a list of first dimension equal to ``depth``, with the :math:`j`-th element of the list
+               corresponding to the value of the argument the :math:`j`-th time the circuit
+               is applied.
+        **kwargs: Static parameters that are passed into ``circuit`` each time it is
+                  repeated (see UsageDetails for more information).
+
+    Raises:
+        ValueError: if inputs do not have the correct format
+
+    .. UsageDetails::
+
+        **Repeating Gates**
+
+        To repeatedly apply a collection of gates/templates, a function applying each of these
+        operations must first be defined. For example, we can define the following circuit:
+
+        .. code-block:: python3
+
+            import pennylane as qml
+            import numpy as np
+
+            def circuit():
+                qml.Hadamard(wires=[0])
+                qml.CNOT(wires=[0, 1])
+                qml.PauliX(wires=[1])
+
+        and then pass it into the ``qml.repeat`` function. In this instance, we repeat ``circuit`` 3 times:
+
+        .. code-block:: python3
+
+            dev = qml.device('default.qubit', wires=3)
+
+            @qml.qnode(dev)
+            def ansatz():
+                qml.repeat(circuit, 3)
+                return [qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))]
+
+
+        This creates the following circuit:
+
+        >>> ansatz()
+        >>> print(ansatz.draw())
+        0: ──H──╭C──X──H──╭C──X──H──╭C──X──┤ ⟨Z⟩
+        1: ─────╰X────────╰X────────╰X─────┤ ⟨Z⟩
+
+        **Static Arguments**
+
+        Static arguments are arguments passed into ``circuit`` that don't change with each
+        repetition of the circuit. Static parameters are always passed as keyword arguments into ``qml.repeat``.
+        For example, consider the following circuit:
+
+        .. code-block:: python3
+
+            def circuit(wires):
+                qml.Hadamard(wires=wires[0])
+                qml.CNOT(wires=wires)
+                qml.PauliX(wires=wires[1])
+
+        We wish to repeat this circuit three times on wires ``1`` and ``2``. Since the wires on which the circuit acts
+        don't change with each repetition of the circuit, the ``wires`` parameter is passed as a keyword argument.
+        We thus repeat the circuit as follows:
```

Comment:
```suggestion
        We therefore repeat the circuit as follows:
```

Lol, I think "thus" is such a science anachronism...maybe better to use "therefore"?

Reply:
Haha, agreed!